### ISSUE TITLE
Multiple Columns

### ISSUE THREAD

**Issue by rocketes** [techprimate/TPPDF#9]
Is it possible to create Text with 2 columns (for example Left / Right) on the same level without using a table?

**Comment by Techprimate** [comment]
Not yet but great idea.

Do you expect it to equally seperate the content, as an example for a paragraph it should draw the text in the first column and the continue in the second column?

How about images? Should 2 images be placed next to each other vertically centered or aligned at the top?

**Comment by rocket-enterprise** [comment]
I would prefer to use it like an questions answer module. So you can set an question on the left side, and the possible answer on the right side (with automatically calculate the needed height for both content rows)

**Comment by Techprimate** [comment]
Sounds like it is probably easier to create by using a table without styling.

If you find a generic way to implement this feature so we can reuse it for other cases to8o, please feel free to Fork & Pull Request your code.

**Comment by marbetschar** [comment]
@techprimate-phil wanted to use the mentioned `PDFMultiSection` since I have to rework some PDFGeneration code and this would come in very handy since I don't know how to solve several requirements without it anyways 🤣 

So I went ahead and checked out the latest and greatest TPPDF version using Carthage - but Xcode does not provide me with any Multisection feature. Tried latest tag and master without success.

How may I use the Multisection feature in my own project?

**Comment by techprimate-phil** [comment]
This was just a proposal and has not been implemented. Do you think this is a good way of handling multi columns?

**Comment by marbetschar** [comment]
@techprimate-phil Ok; I may give an implementation a try, since I need this functionality to get my stuff done. I'd like the version using a section object and would propose the following API as goal - what do you think?

```
let document = PDFDocument(format: .a4)

// init columns using width and alignment
let columns: [PDFSection.Column] = [
    PDFSection.Column(width: 0.33, alignment: .left),
    PDFSection.Column(width: 0.34, alignment: .center),
    PDFSection.Column(width: 0.33, alignment: .right)
]

// init section
let section = PDFSection(columns: columns)

// provide the same api as we have in PDFDocument
// but pass the columnIndex as first argument
section.addText(atColumnIndex: 0, text: "fsfjsljfl")

document.addSection(section)
```

**Comment by techprimate-phil** [comment]
I like this! Better than mine, as it is also using an object oriented approach.
Instead of passing an column index we could use [custom subscripts](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Subscripts.html):

```swift
let section = PDFSection(columns: columns)
section[0].addText(text: "fsfjsljfl")
```

Do you think we should combine it with a per-page multicolumn solution?
You not add it to a specific column, it will put them all in the first column and if the first one is full it will continue in the next one. The result would look like [this](http://standardbaku.club/wp-content/uploads/2017/11/footnote-in-excel-2-tips-to-insert-footnotes-to-multi-column-pages-in-your-word-document-data-recovery-blog-insert-footnote-excel-shortcut.jpg) (found on Google)

I would combine it with the subscript, using `section[0].add...` adds it to a specific section and `section.add...` uses the wrapped solution.

**Comment by marbetschar** [comment]
Subscripting sounds great! I'm currently putting together an initial version (w/o subscripting), hope it works as expected. Would be awesome if you can have a look if the approach fits the expectations.

**Comment by marbetschar** [comment]
@techprimate-phil I created a first attempt:
https://github.com/Mandelkind/TPPDF/tree/section

But I have some sort of calculation error it seems. For some reason, the columns are not aligned vertically on the same height:

![simulator screen shot - iphone 8 plus - 2018-05-05 at 17 41 01](https://user-images.githubusercontent.com/392542/39664936-0851713c-508c-11e8-95dc-0f861fdc66f0.png)


Can you please have a look and point me into the right direction? I can't see what might cause this wrong behaviour.

The code in question I guess is either in the [`PDFSectionObject`](https://github.com/Mandelkind/TPPDF/blob/section/Source/Section/PDFSectionObject.swift) or the [`PDFSectionColumnObject`](https://github.com/Mandelkind/TPPDF/blob/section/Source/Section/PDFSectionColumnObject.swift)

**Comment by marbetschar** [comment]
I think I figured out what the root of the problem is: My code correctly calculates the offset of every column on the page. But for some reason, the generator does not draw the second and third column on the multi column start page.

How may I convince the generator to start drawing every column on the same start page?

**Comment by techprimate-phil** [comment]
I will take a more detailed look and give you feedback, my first thought is probably the unwanted pagebreak. You might have to add the 2. and 3. column before the page break is added, as there is no way to return to previous page when rendering

**Comment by techprimate-phil** [comment]
Why exactly is the following wrapped in  a `if`-statement?

**Comment by marbetschar** [comment]
Copied this from another file. After all I‘m not quite sure I understood how the pdf generation process works in detail. That said, there is for sure room for improvements and I‘m open to suggestions 👍🏼😊

**Comment by marbetschar** [comment]
@techprimate-phil trying to get this done today.

But I'm not quite sure what you mean with "One way would be taking all calculated results of a column and add all objects of all columns on page one to a list and then adding the page break."

Can you please elaborate?

**Comment by marbetschar** [comment]
I think I was able to figure it out. Please note, my multi column addition to the framework does not support multi column text where a long text automatically spawns across the columns.

The way it is implemented right now is primarily for layouting purposes (which is my use case). If someone wants to add support for multi column text I think the best way to do it is to add a flag to `PDFSection` like `isFloatingEnabled` and add the necessary calculations in `PDFSectionObject` as well as in `PDFSectionColumnObject` where needed.

**Comment by techprimate-phil** [comment]
The render engine calculates each object, which returns a list of calculated objects. As an example if you add a text object, it will be calculated and return one text object which will then be rendered. BUT if the text is too long to fit the space, then it will be split up into two text objects with a `PDFPageBreakObject` in-between.

During the render process whenever a page break object is found, it will create a new pdf page and continue there.

In your example when you add the `lorem ipsum` text, it adds a `PDFPageBreakObject` between `...sed diam` and `voluptua...`. 

Now if you add another long text to the second column, it will also break at some point. These need to be merged now! 
the list should now be the following:

- All objects of column 1 before the first pagebreak
- All objects of column 2 before the first pagebreak
- All objects of column 3 before the first pagebreak
- Pagebreak
- All objects of column 1 after the first pagebreak up to the next pagebreak
- All objects of column 2 after the first pagebreak up to the next pagebreak
- All objects of column 3 after the first pagebreak up to the next pagebreak
...

**Comment by techprimate-phil** [comment]
About floating text, when I am going to merge this into the develop branch I will add this functionality by myself.

**Comment by techprimate-phil** [comment]
Okay looks like floating text might need some more work. I will release version 1.1.0 with the current multi section functionality and add it to 1.1.1

**Comment by techprimate-phil** [comment]
Multi-sections are implemented but as of #113 I'll close this now and continue with column wrapping in the other issue
